File: wrappers/nodejs/test/payments.js
const test = require('ava')
const indy = require('../')
const cuid = require('cuid')
const initTestPool = require('./helpers/initTestPool')

test('payments', async function (t) {
  const pool = await initTestPool()

  const walletConfig = { id: 'wallet-' + cuid() }
  const walletCredentials = { key: 'key' }
  await indy.createWallet(walletConfig, walletCredentials)

  const wh = await indy.openWallet(walletConfig, walletCredentials)

  const [trusteeDid] = await indy.createAndStoreMyDid(wh, { seed: '000000000000000000000000Trustee1' })

  const paymentMethod = 'null'
  const paymentAddress = 'pay:null:test'
  const inputs = ['pay:null:1']
  const outputs = [{ recipient: 'pay:null:1', amount: 1 }]
  const from = 1

  let err = await t.throwsAsync(indy.createPaymentAddress(wh, paymentMethod, {}))
  t.is(err.indyName, 'PaymentUnknownMethodError')

  t.deepEqual(await indy.listPaymentAddresses(wh), [])

  err = await t.throwsAsync(indy.addRequestFees(wh, trusteeDid, {}, inputs, outputs, null))
  t.is(err.indyName, 'PaymentUnknownMethodError')

  err = await t.throwsAsync(indy.parseResponseWithFees(paymentMethod, {}))
  t.is(err.indyName, 'PaymentUnknownMethodError')

  err = await t.throwsAsync(indy.buildGetPaymentSourcesRequest(wh, trusteeDid, paymentAddress))
  t.is(err.indyName, 'PaymentUnknownMethodError')

  err = await t.throwsAsync(indy.buildGetPaymentSourcesWithFromRequest(wh, trusteeDid, paymentAddress, from))
  t.is(err.indyName, 'PaymentUnknownMethodError')

  err = await t.throwsAsync(indy.parseGetPaymentSourcesWithFromResponse(paymentMethod, {}))
  t.is(err.indyName, 'PaymentUnknownMethodError')

  err = await t.throwsAsync(indy.buildPaymentReq(wh, trusteeDid, inputs, outputs, null))
  t.is(err.indyName, 'PaymentUnknownMethodError')

  err = await t.throwsAsync(indy.parsePaymentResponse(paymentMethod, {}))
  t.is(err.indyName, 'PaymentUnknownMethodError')

  const extra = await indy.preparePaymentExtraWithAcceptanceData(null, 'indy agreement', '1.0.0', null, 'acceptance mechanism label 1', 123379200)
  const expectedExtra = {
    mechanism: 'acceptance mechanism label 1',
    taaDigest: '7213b9aabf8677edf6b17d20a9fbfaddb059ea4cb122d163bdf658ea67196120',
    time: 123379200
  }
  t.deepEqual(extra.taaAcceptance, expectedExtra)

  err = await t.throwsAsync(indy.buildMintReq(wh, trusteeDid, outputs, null))

  t.is(err.indyName, 'PaymentUnknownMethodError')

  let fees = { txnType1: 1, txnType2: 2 }

  err = await t.throwsAsync(indy.buildSetTxnFeesReq(wh, trusteeDid, paymentMethod, fees))
  t.is(err.indyName, 'PaymentUnknownMethodError')

  err = await t.throwsAsync(indy.buildGetTxnFeesReq(wh, trusteeDid, paymentMethod))
  t.is(err.indyName, 'PaymentUnknownMethodError')

  err = await t.throwsAsync(indy.parseGetTxnFeesResponse(paymentMethod, {}))
  t.is(err.indyName, 'PaymentUnknownMethodError')

  const receipt = 'pay:null:0_PqVjwJC42sxCTJp'

  err = await t.throwsAsync(indy.buildVerifyPaymentReq(wh, trusteeDid, receipt))
  t.is(err.indyName, 'PaymentUnknownMethodError')

  err = await t.throwsAsync(indy.parseVerifyPaymentResponse(paymentMethod, {}))
  t.is(err.indyName, 'PaymentUnknownMethodError')

  const message = Buffer.from('123456789', 'utf8')
  err = await t.throwsAsync(indy.signWithAddress(wh, paymentAddress, message))
  t.is(err.indyName, 'PaymentUnknownMethodError')

  const signature = Buffer.from('987654321', 'utf8')
  err = await t.throwsAsync(indy.verifyWithAddress(paymentAddress, message, signature))
  t.is(err.indyName, 'PaymentUnknownMethodError')

  const getAuthRuleResp = { result: { data: [{ new_value: '0', constraint: { need_to_be_owner: false, sig_count: 1, metadata: { fees: '1' }, role: '0', constraint_id: 'ROLE' }, field: 'role', auth_type: '1', auth_action: 'ADD' }], identifier: 'LibindyDid111111111111', auth_action: 'ADD', new_value: '0', reqId: 15616, auth_type: '1', type: '121', field: 'role' }, op: 'REPLY' }
  const requesterInfo = { role: '0', need_to_be_owner: false, sig_count: 1 }
  fees = { 1: 100 }

  const requestInfo = await indy.getRequestInfo(getAuthRuleResp, requesterInfo, fees)
  const expectedRequestInfo = {
    price: 100,
    requirements: [{
      role: '0',
      need_to_be_owner: false,
      sig_count: 1
    }]
  }
  t.deepEqual(expectedRequestInfo, requestInfo)

  await indy.closeWallet(wh)
  await indy.deleteWallet(walletConfig, walletCredentials)

  pool.cleanup()
})

File: libnullpay/tests/utils/payments.rs
use indy::IndyError;
use indy::payments;
use indy::future::Future;
use indy::WalletHandle;

pub fn create_payment_address(wallet_handle: WalletHandle, payment_method: &str, config: &str) -> Result<String, IndyError> {
    payments::create_payment_address(wallet_handle, payment_method, config).wait()
}

pub fn list_payment_addresses(wallet_handle: WalletHandle) -> Result<String, IndyError> {
    payments::list_payment_addresses(wallet_handle).wait()
}

pub fn add_request_fees(wallet_handle: WalletHandle, submitter_did: &str, req_json: &str, inputs_json: &str, outputs_json: &str, extra: Option<&str>) -> Result<(String, String), IndyError> {
    payments::add_request_fees(wallet_handle, Some(submitter_did), req_json, inputs_json, outputs_json, extra).wait()
}

pub fn parse_response_with_fees(payment_method: &str, resp_json: &str) -> Result<String, IndyError> {
    payments::parse_response_with_fees(payment_method, resp_json).wait()
}

pub fn build_get_payment_sources_request(wallet_handle: WalletHandle, submitter_did: &str, payment_address: &str) -> Result<(String, String), IndyError> {
    payments::build_get_payment_sources_request(wallet_handle, Some(submitter_did), payment_address).wait()
}


pub fn parse_get_payment_sources_response(payment_method: &str, resp_json: &str) -> Result<String, IndyError> {
    payments::parse_get_payment_sources_response(payment_method, resp_json).wait()
}

pub fn build_payment_req(wallet_handle: WalletHandle, submitter_did: &str, inputs: &str, outputs: &str, extra: Option<&str>) -> Result<(String, String), IndyError> {
    payments::build_payment_req(wallet_handle, Some(submitter_did), inputs, outputs, extra).wait()
}

pub fn parse_payment_response(payment_method: &str, resp_json: &str) -> Result<String, IndyError> {
    payments::parse_payment_response(payment_method, resp_json).wait()
}

pub fn build_mint_req(wallet_handle: WalletHandle, submitter_did: &str, outputs_json: &str, extra: Option<&str>) -> Result<(String, String), IndyError> {
    payments::build_mint_req(wallet_handle, Some(submitter_did), outputs_json, extra).wait()
}

pub fn build_set_txn_fees_req(wallet_handle: WalletHandle, submitter_did: &str, payment_method: &str, fees_json: &str) -> Result<String, IndyError> {
    payments::build_set_txn_fees_req(wallet_handle, Some(submitter_did), payment_method, fees_json).wait()
}

pub fn build_get_txn_fees_req(wallet_handle: WalletHandle, submitter_did: &str, payment_method: &str) -> Result<String, IndyError> {
    payments::build_get_txn_fees_req(wallet_handle, Some(submitter_did), payment_method).wait()
}

pub fn parse_get_txn_fees_response(payment_method: &str, resp_json: &str) -> Result<String, IndyError> {
    payments::parse_get_txn_fees_response(payment_method, resp_json).wait()
}

pub fn build_verify_payment_req(wallet_handle: WalletHandle, submitter_did: &str, receipt: &str) -> Result<(String, String), IndyError> {
    payments::build_verify_payment_req(wallet_handle, Some(submitter_did), receipt).wait()
}

pub fn parse_verify_payment_response(payment_method: &str, resp_json: &str) -> Result<String, IndyError> {
    payments::parse_verify_payment_response(payment_method, resp_json).wait()
}

pub fn sign_with_address(wallet_handle: WalletHandle, address: &str, message: &[u8]) -> Result<Vec<u8>, IndyError> {
    payments::sign_with_address(wallet_handle, address, message).wait()
}

pub fn verify_with_address(address: &str, message: &[u8], signature: &[u8]) -> Result<bool, IndyError> {
    payments::verify_with_address(address, message, signature).wait()
}
